### PR TITLE
Get rid of `do { ... } catch { ... }` for expected errors

### PR DESCRIPTION
**Pull request by shekhar-rajak**
### Motivation:

Get rid of `do { ... } catch { ... }` for expected errors

### Modifications:

Replacing all uses of this pattern in Tests/**.

```
do {
    try someOperation()
XCTFail("should throw") // easy to forget
} catch error as SomethingError {
    XCTAssertEqual(.something, error as? SomethingError)
} catch {
    XCTFail("wrong error")
}
```

with 

```
XCTAssertThrowsError(try someOperation) { error in
    XCTAssertEqual(.something, error as? SomethingError)
}
```

### Result:

Fixes https://github.com/apple/swift-nio-extras/issues/83

### CODE REVIEW

File: Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift [hunk 1]
```diff
@@ -194,23 +194,22 @@ class LineBasedFrameDecoderTest: XCTestCase {
             buffer.writeString(string)
             return buffer
         }
-
+
```
Comment (shekhar-rajak) [inline Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift]: Please suggest me a plugin to remove trailing whitespaces from file while saving the file.

File: Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift [hunk 2]
```diff
@@ -65,12 +65,10 @@ class HTTPRequestDecompressorTest: XCTestCase {
 
         do {
```
Comment (shekhar-rajak) [inline Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: Instead of this do catch block I was trying `XCTAssertThrowsSpecific` but it was not working: 

```
error: use of unresolved identifier 'XCTAssertThrowsSpecific'
```

File: Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift [hunk 3]
```diff
@@ -205,12 +205,13 @@ class LineBasedFrameDecoderTest: XCTestCase {
...
                                                                      byteBuffer("")])
-            ]) {
+                ]){
                 return LineBasedFrameDecoder()
             }
+
         } catch {
             print(error)
-            XCTFail()
+            XCTFail("Unexptected error: \(error)")
```
Comment (Lukasa) [inline Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift]:
```suggestion
            XCTFail("Unexpected error: \(error)")
```

File: Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift [hunk 4]
```diff
@@ -205,12 +205,13 @@ class LineBasedFrameDecoderTest: XCTestCase {
                 ("Content-Length: 17\r\nConnection: close\r\n\r\n", [byteBuffer("Content-Length: 17"),
                                                                      byteBuffer("Connection: close"),
                                                                      byteBuffer("")])
-            ]) {
+                ]){
```
Comment (Lukasa) [inline Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift]: Remove this whitespace change.

File: Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift [hunk 5]
```diff
@@ -205,12 +205,13 @@ class LineBasedFrameDecoderTest: XCTestCase {
...
                 return LineBasedFrameDecoder()
             }
+
```
Comment (Lukasa) [inline Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift]: Remove this whitespace change.

File: Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift (hunk 2 above)
Comment (Lukasa) [reply Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: That function is for exceptions only: you should use `XCTAssertThrowsError`.

File: Tests/NIOHTTPCompressionTests/HTTPResponseCompressorTest.swift [hunk 6]
```diff
@@ -525,10 +525,8 @@ class HTTPResponseCompressorTest: XCTestCase {
         }
 
         XCTAssertNoThrow(try channel.pipeline.removeHandler(name: "compressor").wait())
-
-        do {
-            try writePromise.futureResult.wait()
-        } catch {
+        
+        XCTAssertThrowsError(try writePromise.futureResult.wait()) {error in
```
Comment (Lukasa) [inline Tests/NIOHTTPCompressionTests/HTTPResponseCompressorTest.swift]: This change is not quite right: we don't _know_ that this will throw an error, nor do we assert it must, it just _might_ and we want to swallow that. Please revert.

File: Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift (hunk 2 above)
Comment (shekhar-rajak) [reply Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: Actually I was getting error with AssetEqual: ` error: generic parameter 'T' could not be inferred`

File: Tests/NIOHTTPCompressionTests/HTTPResponseCompressorTest.swift (hunk 6 above)
Comment (shekhar-rajak) [reply Tests/NIOHTTPCompressionTests/HTTPResponseCompressorTest.swift]: Okay I have reverted this change.

File: Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift [hunk 7]
```diff
@@ -205,12 +205,12 @@ class LineBasedFrameDecoderTest: XCTestCase {
                 ("Content-Length: 17\r\nConnection: close\r\n\r\n", [byteBuffer("Content-Length: 17"),
                                                                      byteBuffer("Connection: close"),
                                                                      byteBuffer("")])
-            ]) {
+            ]){
```
Comment (Lukasa) [inline Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift]: There's still a whitespace change here.

File: Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift (hunk 2 above)
Comment (Lukasa) [reply Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: `DecompressionError` is not `Equatable`, so that could not work. It's unrelated to the use of `XCTAssertThrowsError` though: it only affects the code in the `error` block.

File: Tests/NIOHTTPCompressionTests/HTTPResponseCompressorTest.swift [hunk 8]
```diff
@@ -525,7 +525,7 @@ class HTTPResponseCompressorTest: XCTestCase {
         }
 
         XCTAssertNoThrow(try channel.pipeline.removeHandler(name: "compressor").wait())
-
+
```
Comment (Lukasa) [inline Tests/NIOHTTPCompressionTests/HTTPResponseCompressorTest.swift]: Can we remove this whitespace change?

File: Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift (hunk 2 above)
Comment (shekhar-rajak) [reply Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: So, should I revert  the related changes (`if case .limit = `) in all the file ?

File: Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift (hunk 7 above)
Comment (shekhar-rajak) [reply Tests/NIOExtrasTests/LineBasedFrameDecoderTest.swift]: Done!

File: Tests/NIOHTTPCompressionTests/HTTPResponseCompressorTest.swift (hunk 8 above)
Comment (shekhar-rajak) [reply Tests/NIOHTTPCompressionTests/HTTPResponseCompressorTest.swift]: Done!

File: Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift (hunk 2 above)
Comment (Lukasa) [reply Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: Those changes are fine, but they need to be moved into the closure that `XCTAssertThrowsError` provides.

Comment (shekhar-rajak) [reply Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: I think last time I tested there were error : 

```
HTTPResponseDecompressorTest.swift:91: error: -[NIOHTTPCompressionTests.HTTPResponseDecompressorTest testDecompression] : XCTAssertThrowsError failed: did not throw an error
```

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

File: Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift [hunk 9]
```diff
@@ -63,14 +63,10 @@ class HTTPRequestDecompressorTest: XCTestCase {
...
-        do {
-            try channel.writeInbound(HTTPServerRequestPart.body(compressed))
-        } catch let error as NIOHTTPDecompression.DecompressionError {
-            switch error {
-            case .limit:
-                // ok
-                break
-            default:
+        XCTAssertThrowsError(try channel.writeInbound(HTTPServerRequestPart.body(compressed))) {error in
+           if case .limit = error as? NIOHTTPDecompression.DecompressionError {
```
Comment (Lukasa) [inline Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: For Swift 5.0 these all need to be `if case .some(.limit)`

Comment (shekhar-rajak) [reply Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: Thanks, I did the change

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

File: Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift [hunk 10]
```diff
@@ -63,14 +63,10 @@ class HTTPRequestDecompressorTest: XCTestCase {
...
-        } catch let error as NIOHTTPDecompression.DecompressionError {
-            switch error {
-            case .limit:
-                // ok
-                break
-            default:
+        XCTAssertThrowsError(try channel.writeInbound(HTTPServerRequestPart.body(compressed))) {error in
+           if case .some(.limit) = error as? NIOHTTPDecompression.DecompressionError {
+                // Okay
+            } else {
```
Comment (Lukasa) [inline Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: Ok, let's revert this change entirely: as you note, the assertion is not valid here, so we shouldn't bother to make the change.

File: Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift [hunk 11]
```diff
@@ -120,14 +112,10 @@ class HTTPRequestDecompressorTest: XCTestCase {
...
-            } catch let error as NIOHTTPDecompression.DecompressionError {
-                switch error {
-                case .limit:
-                    // ok
-                    break
-                default:
+            XCTAssertThrowsError(try channel.writeInbound(HTTPServerRequestPart.body(compressed))) { error in
+               if case .some(.limit) = error as? NIOHTTPDecompression.DecompressionError {
+                    // Okay
+                } else {
```
Comment (Lukasa) [inline Tests/NIOHTTPCompressionTests/HTTPRequestDecompressorTest.swift]: And similarly let's totally revert this change, as the assertion isn't valid here either.

File: Tests/NIOHTTPCompressionTests/HTTPResponseDecompressorTest.swift [hunk 12]
```diff
@@ -98,18 +88,12 @@ class HTTPResponseDecompressorTest: XCTestCase {
...
-            } catch let error as NIOHTTPDecompression.DecompressionError {
-                switch error {
-                case .limit:
-                    // ok
-                    break
-                default:
+            XCTAssertThrowsError(try channel.writeInbound(HTTPClientResponsePart.body(compressed))) { error in
+                if case .some(.limit) = error as? NIOHTTPDecompression.DecompressionError {
+                    // Okay
+                } else {
```
Comment (Lukasa) [inline Tests/NIOHTTPCompressionTests/HTTPResponseDecompressorTest.swift]: And similarly let's totally revert this change too.

**Comment by shekhar-rajak** [comment]
I have reverted the changes, which were failing.

**Comment by Lukasa** [comment]
@swift-nio-bot test this please
